src/Simulation/Core/Operations/Controlled.cs

```diff
@@ -110,6 +113,25 @@ public ControlledOperation(Operation<I, O> op) : base(op.Factory)
 
         public override IApplyData __dataOut(QVoid data) => data;
 
+        /// <inheritdoc/>
+        public override RuntimeMetadata? GetRuntimeMetadata(IApplyData args)
+        {
+            if (args.Value is ValueTuple<IQArray<Qubit>, I> ctrlArgs)
+            {
+                var (controls, baseArgs) = ctrlArgs;
+                var baseMetadata = this.BaseOp.GetRuntimeMetadata(this.BaseOp.__dataIn(baseArgs));
+                if (baseMetadata == null) return null;
+                baseMetadata.IsControlled = true;
+                baseMetadata.Controls = controls.Concat(baseMetadata.Controls);
+                return baseMetadata;
+            }
+            else
+            {
+                Console.WriteLine($"Failed to retrieve runtime metadata for {this.ToString()}.");
```

Comment:
This should really be a `Logger`.
Maybe instead, make a `Debug.Assert` for the condition?

Reply:
Would that look like a `Debug.Assert(false, ...)` or more like `Debug.Assert(args.Value is ValueTuple<IQArray<Qubit>, I>)`? I tried the latter but it got pretty ugly with handling `null`s after the assert statement (i.e. have to handle the case when `args.Value` is not of the required type even though the `assert` statement should've caught it).